state/protocol/inmem/epoch.go

```diff
@@ -23,30 +23,32 @@ func (eq Epochs) Previous() protocol.Epoch {
 	if eq.entry.PreviousEpoch == nil {
 		return invalid.NewEpoch(protocol.ErrNoPreviousEpoch)
 	}
-	return NewCommittedEpoch(eq.entry.PreviousEpochSetup, eq.entry.PreviousEpochCommit)
+	return NewCommittedEpoch(eq.entry.PreviousEpochSetup, eq.entry.PreviousEpoch.EpochExtensions, eq.entry.PreviousEpochCommit)
 }
 
 func (eq Epochs) Current() protocol.Epoch {
-	return NewCommittedEpoch(eq.entry.CurrentEpochSetup, eq.entry.CurrentEpochCommit)
+	return NewCommittedEpoch(eq.entry.CurrentEpochSetup, eq.entry.CurrentEpoch.EpochExtensions, eq.entry.CurrentEpochCommit)
 }
 
 func (eq Epochs) Next() protocol.Epoch {
 	switch eq.entry.EpochPhase() {
 	case flow.EpochPhaseStaking, flow.EpochPhaseFallback:
 		return invalid.NewEpoch(protocol.ErrNextEpochNotSetup)
 	case flow.EpochPhaseSetup:
-		return NewSetupEpoch(eq.entry.NextEpochSetup)
+		return NewSetupEpoch(eq.entry.NextEpochSetup, eq.entry.NextEpoch.EpochExtensions)
 	case flow.EpochPhaseCommitted:
-		return NewCommittedEpoch(eq.entry.NextEpochSetup, eq.entry.NextEpochCommit)
+		return NewCommittedEpoch(eq.entry.NextEpochSetup, eq.entry.NextEpoch.EpochExtensions, eq.entry.NextEpochCommit)
 	}
 	return invalid.NewEpochf("unexpected unknown phase in protocol state entry")
 }
 
-// setupEpoch is an implementation of protocol.Epoch backed by an EpochSetup
-// service event. This is used for converting service events to inmem.Epoch.
+// setupEpoch is an implementation of protocol.Epoch backed by an EpochSetup service event.
+// Includes any extensions which have been included as of the reference block.
+// This is used for converting service events to inmem.Epoch.
 type setupEpoch struct {
 	// EpochSetup service event
 	setupEvent *flow.EpochSetup
+	extensions []flow.EpochExtension
 }
```

Comment:
> only committed epochs can have extensions

👍 Agree, will move the extensions to `committedEpoch`.

> We have the convention that an epoch essentially only starts to exist after the EpochCommit Service event has been finalized

Epoch APIs cannot strictly work this way, because the node software requires access to a subset of un-committed epoch data in order to commit that epoch, by completing the DKG and generating root cluster QCs. 

There are other ways we can improve the epoch APIs, but we will need something similar to the progressive disclosure of `setupEpoch` and `committedEpoch`. For example, we could hide some info from the EpochSetup event until the epoch is committed, and we could use different interface types to represent the two states.

Reply:
Added this tech debt issue as a place to capture these ideas for improving the API: https://github.com/onflow/flow-go/issues/6191